File: Modify Linked List-1 - GFG/modify-linked-list1.cpp
//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;
struct Node* modifyTheList(struct Node *head);
struct Node
{
    int data;
    struct Node* next;
    
    Node(int x){
        data = x;
        next = NULL;
    }
    
};

void print(Node *head)
{
    Node *temp=head;
	while (temp)
    {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout<<endl;
}


// } Driver Code Ends
/*Complete the function below
Node is as follows:
struct Node
{
    int data;
    struct Node* next;
    
    Node(int x){
        data = x;
        next = NULL;
    }
    
};
*/

class Solution{
    public:
    struct Node* modifyTheList(struct Node *head)
    {
     Node* mid = get_m(head);

    // Partition 2nd list
    Node* backlist = mid->next; // 3->6->NULL

    // Mark next as null for the last node of the first list
    mid->next = nullptr; // 10>4>5>NULL

    backlist = reverse(backlist); // 6>3>NULL

    // Iterate in both lists, store the difference in the 1st and copy 1st list data to 2nd
    Node* curr1 = head;
    Node* curr2 = backlist;
    while (curr2 != nullptr) {
        int temp = curr1->data;
        curr1->data = curr2->data - temp;
        curr2->data = temp;

        curr1 = curr1->next;
        curr2 = curr2->next;
    }
    // 1st: -4, -1, 5->NULL || 2nd: 10>4>NULL
    // Reverse 2nd list and append it to the 1st (last node of the 1st list is mid)
    mid->next = reverse(backlist);
    return head;
}

Node* get_m(Node* head) {
    Node* f = head;
    Node* s = head;
    while (f->next != nullptr && f->next->next != nullptr) {
        s = s->next;
        f = f->next->next;
    }
    return s;
}

Node* reverse(Node* head) {
    Node* prev = nullptr;
    Node* curr = head;
    Node* next = nullptr;
    while (curr != nullptr) {
        next = curr->next;
        curr->next = prev;
        prev = curr;

        curr = next;
    }
    return prev;
}
};

//{ Driver Code Starts.
int main()
{
	int t;
	cin>>t;
	while(t--){
		int n;
		cin>>n;
		struct Node *head = NULL,*temp=NULL;
		while(n--){
		    int a;
            cin>>a;
			Node *newNode = new Node(a);
			if(head==NULL)
                head=newNode;
            else
                temp->next=newNode;
            temp=newNode;
		}
		Solution obj;
		head = obj.modifyTheList(head);
		print(head);
		
	}
    return 0;
}
// } Driver Code Ends

File: Modify Linked List-1 - GFG/modify-linked-list1.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class Node
{
    int data;
    Node next;
    public Node(int data)
    {
        this.data=data;
    }
}
class GFG{
    public static void main(String args[]) throws IOException { 
        BufferedReader in=new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out=new PrintWriter(System.out);
        int t = Integer.parseInt(in.readLine());

        while (t-- > 0) {
            int n=Integer.parseInt(in.readLine().trim());
            String s[]=in.readLine().trim().split(" ");
            Node head=new Node(Integer.parseInt(s[0]));
            Node copy=head;
            for(int i=1;i<n;i++){
                Node temp=new Node(Integer.parseInt(s[i]));
                copy.next=temp;
                copy=copy.next;
            }
            Solution ob=new Solution();
            Node ans=ob.modifyTheList(head);
            while(ans!=null){
                out.print(ans.data+" ");
                ans=ans.next;
            }out.println();
        }
        out.close();
    }
}
// } Driver Code Ends


//User function Template for Java
class Solution
{
    public static Node modifyTheList(Node head)
    {
        //make array and ele in list
        ArrayList<Integer> al = new ArrayList<>();
        Node curr = head;
        while(curr!=null)
        {
            al.add(curr.data);
            curr=curr.next;
        }
        int n = al.size();
         curr=head;
        for(int i=0; i<n/2; i++)
        {
            int x =   al.get(n-i-1) - al.get(i);
           al.set(n-i-1,al.get(i));
           
           al.set(i,x);
        }
        int i =0; 
        while(curr!=null)
        {
            curr.data = al.get(i++);
            curr=curr.next;
        }
        return head;
    }
}
